### PR TITLE
LPS-153319 Allow to set/update extended properties in existing Liferay Portal Entities

### PR DESCRIPTION
**Pull request by javierdearcos**
This PR defines the new API Extensibility Framework that will allow the extension of System Objects to set/update extended properties and get these extended properties.

The new interface is called **ExtensionProvider** and is meant to replace the old _ExtensionContext_ mechanism. The definition of the new interface is the following:
```
public interface ExtensionProvider {

        // To be called to get the extended properties of an object
	public Map<String, Serializable> getExtendedProperties(
			long companyId, Object entity)
		throws Exception;

         // To be called to get the definition of the extended properties to validate and to document
	public Map<String, ExtendedPropertyDefinition>
		getExtendedPropertyDefinitions(long companyId, String className);

         // To be called to get the filtered properties of an object
	public Collection<String> getFilteredPropertyNames(
		long companyId, Object entity);

        // To check if the extension applies to the given company and className. Used by the infrastructure to select the correct ExtensionProviders
	public boolean isExtensionFor(long companyId, String className);

        // To be called to set/update the extended properties of an object
	public void setExtendedProperties(
			long companyId, Object entity,
			Map<String, Serializable> extensionProviderExtendedProperties)
		throws Exception;

}
```
The infrastructure includes an **ExtensionProviderRegistry** that will register all ExtensionProviders active in the portal. This registry is used by the **ExtensionProvidersContextResolver** that will be called by the infrastructure to get the ExtensionProviders class, which contains the list of all ExtensionProviders that applies to the object in the request.

The **ExtensionProviders** class is necessary to have a ContextResolver that will inject this class, and also ease the aggregation logic by calling all extension providers that apply to a given object.

The **EntityExtensionThreadLocal** is needed to save the extended properties that come in the message body and then recover them in the writer interceptors to call the set/update logic.

The workflow is:
1. A new request is received
2. The JAX-RS message body reader checks if the class of the message body has any ExtensionProvider
3. If true get the extended property definitions to validate the extended properties received in the message body
4. Save the extended properties in the thread-local to be recovered later
5. Process the original request
6. After processing the request, in the JAX-RS writer interceptors recover the extended properties
7. If exists call the setExtendedProperties to set/update the extended property values
8. Call the getExtendedProperties to be included in the response
9. Proceed to send the response

This is the first PR to start establishing the infrastructure

### CODE REVIEW

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/message/body/BaseMessageBodyReader.java [hunk 1]
```diff
@@ -94,6 +152,35 @@ private ObjectMapper _getObjectMapper(Class<?> clazz) {
...
+			if (!clazzFieldNames.contains(jsonFieldName)) {
+				JsonNode fieldNameJsonNode = jsonNode.get(jsonFieldName);
+
+				extendedProperties.put(
+					jsonFieldName,
+					_getJsonNodeValue(fieldNameJsonNode, objectMapper));
+			}
+		}
+
+		extensionContext.setExtendedProperties(object, extendedProperties);
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/message/body/BaseMessageBodyReader.java:192]: Is necessary to call `setExtendedProperties` even if there are no properties to set?

Comment (javierdearcos) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/message/body/BaseMessageBodyReader.java:192]: Good point.

It was unintentional but thinking about it now, maybe it is a good idea to allow setting default values or making some checks.

I would leave it like that and maybe clarify in the usage that it is meant for this purpose.

**Comment by javierdearcos** [comment]
Today I will work a little more to check how we can add a transaction to be able to rollback the complete object if there is any problem and make more tests of the extension and the error handling

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/message/body/BaseMessageBodyReader.java (hunk 1 above)
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/message/body/BaseMessageBodyReader.java:192]: I like the idea. One question, where is documented the usage? In the Pull request / Jira issue?

Comment (javierdearcos) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/message/body/BaseMessageBodyReader.java:192]: We will need to document it after the final version is merged. Should be one task of the epic (the last one) and in this kind of epic is very important to document it as would be an "important" feature

**Comment by javierdearcos** [comment]
!()[https://media.giphy.com/media/ujUdrdpX7Ok5W/giphy.gif]

**Comment by javierdearcos** [comment]
ci:test:sf

**Comment by javierdearcos** [comment]
ci:test:relevant

**Comment by javierdearcos** [comment]
ci:test:sf

**Comment by javierdearcos** [comment]
ci:test:relevant

**Review by ccorreagg — APPROVED** [review]
Great job!

**Comment by ccorreagg** [comment]
ci:forward

**Pull request closed without merging**